File: packages/sdk-node/src/core/focus-export.ts
/**
 * FOCUS export — spend data in the FinOps interchange shape.
 *
 * FOCUS (FinOps Open Cost and Usage Specification, focus.finops.org) is
 * the column contract enterprise cost tooling ingests. This export is
 * FOCUS-ALIGNED: core columns follow the spec (ConsumedQuantity/
 * ConsumedUnit are the spec's own home for token usage — FOCUS 1.2's
 * virtual-currency work uses GPT tokens as its worked example), and
 * RateGuard-specific detail rides in x_-prefixed columns, the spec's
 * sanctioned extension convention.
 *
 * Honest scope: costs here are RateGuard's pricing-table ESTIMATES of
 * LLM spend observed in-process, not a provider invoice. BilledCost is
 * deliberately 0 — RateGuard bills nothing; EffectiveCost carries the
 * estimate. Reconcile against provider billing for accounting truth.
 */

import type { SpendReceipt } from './spend-receipt.js';

export const FOCUS_HEADER = [
  'ChargePeriodStart', 'ChargePeriodEnd', 'ChargeCategory', 'ChargeDescription',
  'BilledCost', 'EffectiveCost', 'BillingCurrency',
  'ProviderName', 'ServiceName', 'ServiceCategory',
  'ResourceId', 'SkuId', 'ConsumedQuantity', 'ConsumedUnit',
  'x_rateguard_input_tokens', 'x_rateguard_output_tokens',
  'x_rateguard_policy_preset', 'x_rateguard_attestation_token_id',
  'x_rateguard_receipt_signature',
] as const;

/** One charge-period row; property order is the CSV column order. */
export interface FOCUSRow {
  chargePeriodStart: string;
  chargePeriodEnd: string;
  chargeCategory: string;
  chargeDescription: string;
  billedCost: number;
  effectiveCost: number;
  billingCurrency: string;
  providerName: string;
  serviceName: string;
  serviceCategory: string;
  resourceId: string;
  skuId: string;
  consumedQuantity: number;
  consumedUnit: string;
  xInputTokens: number;
  xOutputTokens: number;
  xPolicyPreset: string;
  xAttestationTokenId: string;
  xReceiptSignature: string;
}

function isoUTC(unixSeconds: number): string {
  return new Date(unixSeconds * 1000).toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/** Maps a spend receipt onto a FOCUS row. */
export function focusRowFromReceipt(r: SpendReceipt): FOCUSRow {
  const model = r.claims.model ?? '';
  const desc = model
    ? `LLM token usage (${model}) metered in-process by RateGuard`
    : 'LLM token usage metered in-process by RateGuard';
  return {
    chargePeriodStart: isoUTC(r.claims.windowStartUnix),
    chargePeriodEnd: isoUTC(r.claims.windowEndUnix),
    chargeCategory: 'Usage',
    chargeDescription: desc,
    billedCost: 0,
    effectiveCost: r.claims.estimatedCostMicroUSD / 1e6,
    billingCurrency: 'USD',
    providerName: r.claims.provider ?? '',
    serviceName: 'LLM Inference',
    serviceCategory: 'AI and Machine Learning',
    resourceId: r.claims.key,
    skuId: model,
    consumedQuantity: r.claims.totalTokens,
    consumedUnit: 'tokens',
    xInputTokens: r.claims.inputTokens,
    xOutputTokens: r.claims.outputTokens,
    xPolicyPreset: r.claims.policyPreset ?? '',
    xAttestationTokenId: r.claims.attestationTokenId ?? '',
    xReceiptSignature: r.signature.length > 0 ? r.signature.toString('base64') : '',
  };
}

/**
 * Minimal RFC-4180 escaping, matching Go's encoding/csv output for these
 * cells: quote only when the cell contains a comma, quote, or newline.
 */
function csvCell(value: string): string {
  if (/[",\n\r]/.test(value)) {
    return `"${value.split('"').join('""')}"`;
  }
  return value;
}

/**
 * Formats numbers like Go's %g so all three SDKs emit identical cells
 * (0.645 → "0.645", 154500 → "154500", 0 → "0").
 */
function fmt(value: number): string {
  return String(value);
}

/** Renders header plus one line per row, \n-terminated (Go's csv default). */
export function writeFOCUSCSV(rows: FOCUSRow[]): string {
  const lines = [FOCUS_HEADER.join(',')];
  for (const r of rows) {
    lines.push(
      [
        r.chargePeriodStart, r.chargePeriodEnd, r.chargeCategory, r.chargeDescription,
        fmt(r.billedCost), fmt(r.effectiveCost), r.billingCurrency,
        r.providerName, r.serviceName, r.serviceCategory,
        r.resourceId, r.skuId, fmt(r.consumedQuantity), r.consumedUnit,
        String(r.xInputTokens), String(r.xOutputTokens),
        r.xPolicyPreset, r.xAttestationTokenId, r.xReceiptSignature,
      ]
        .map(csvCell)
        .join(','),
    );
  }
  return lines.join('\n') + '\n';
}

File: packages/sdk-python/rateguard/core/focus_export.py
"""FOCUS export — spend data in the FinOps interchange shape.

FOCUS (FinOps Open Cost and Usage Specification, focus.finops.org) is the
column contract enterprise cost tooling ingests. This export is
FOCUS-ALIGNED: core columns follow the spec (ConsumedQuantity/ConsumedUnit
are the spec's own home for token usage — FOCUS 1.2's virtual-currency
work uses GPT tokens as its worked example), and RateGuard-specific
detail rides in ``x_``-prefixed columns, the spec's sanctioned extension
convention.

Honest scope: costs here are RateGuard's pricing-table ESTIMATES of LLM
spend observed in-process, not a provider invoice. BilledCost is
deliberately 0 — RateGuard bills nothing; EffectiveCost carries the
estimate. Reconcile against provider billing for accounting truth.
"""

from __future__ import annotations

import base64
import csv
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import IO

from .spend_receipt import SpendReceipt

FOCUS_HEADER = [
    "ChargePeriodStart", "ChargePeriodEnd", "ChargeCategory", "ChargeDescription",
    "BilledCost", "EffectiveCost", "BillingCurrency",
    "ProviderName", "ServiceName", "ServiceCategory",
    "ResourceId", "SkuId", "ConsumedQuantity", "ConsumedUnit",
    "x_rateguard_input_tokens", "x_rateguard_output_tokens",
    "x_rateguard_policy_preset", "x_rateguard_attestation_token_id",
    "x_rateguard_receipt_signature",
]


@dataclass(frozen=True)
class FOCUSRow:
    """One charge-period row; field order is the CSV column order."""

    charge_period_start: str
    charge_period_end: str
    charge_category: str
    charge_description: str
    billed_cost: float
    effective_cost: float
    billing_currency: str
    provider_name: str
    service_name: str
    service_category: str
    resource_id: str
    sku_id: str
    consumed_quantity: float
    consumed_unit: str
    x_input_tokens: int
    x_output_tokens: int
    x_policy_preset: str
    x_attestation_token_id: str
    x_receipt_signature: str


def _iso_utc(unix_seconds: int) -> str:
    return datetime.fromtimestamp(unix_seconds, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def focus_row_from_receipt(receipt: SpendReceipt) -> FOCUSRow:
    c = receipt.claims
    desc = "LLM token usage metered in-process by RateGuard"
    if c.model:
        desc = f"LLM token usage ({c.model}) metered in-process by RateGuard"
    return FOCUSRow(
        charge_period_start=_iso_utc(c.window_start_unix),
        charge_period_end=_iso_utc(c.window_end_unix),
        charge_category="Usage",
        charge_description=desc,
        billed_cost=0.0,
        effective_cost=c.estimated_cost_micro_usd / 1e6,
        billing_currency="USD",
        provider_name=c.provider,
        service_name="LLM Inference",
        service_category="AI and Machine Learning",
        resource_id=c.key,
        sku_id=c.model,
        consumed_quantity=float(c.total_tokens),
        consumed_unit="tokens",
        x_input_tokens=c.input_tokens,
        x_output_tokens=c.output_tokens,
        x_policy_preset=c.policy_preset,
        x_attestation_token_id=c.attestation_token_id,
        x_receipt_signature=base64.b64encode(receipt.signature).decode() if receipt.signature else "",
    )


def _fmt(value: float) -> str:
    # Mirrors Go's %g so the three SDKs emit identical CSV cells.
    return f"{value:g}"


def write_focus_csv(stream: IO[str], rows: list[FOCUSRow]) -> None:
    """Write a header plus one line per row."""
    # lineterminator matches Go's encoding/csv default (\n) so all three
    # SDKs emit byte-identical exports for identical rows.
    writer = csv.writer(stream, lineterminator="\n")
    writer.writerow(FOCUS_HEADER)
    for r in rows:
        writer.writerow(
            [
                r.charge_period_start, r.charge_period_end, r.charge_category, r.charge_description,
                _fmt(r.billed_cost), _fmt(r.effective_cost), r.billing_currency,
                r.provider_name, r.service_name, r.service_category,
                r.resource_id, r.sku_id, _fmt(r.consumed_quantity), r.consumed_unit,
                str(r.x_input_tokens), str(r.x_output_tokens),
                r.x_policy_preset, r.x_attestation_token_id,
                r.x_receipt_signature,
            ]
        )
